src/Http/Http.Extensions/src/RequestDelegateFactory.cs

```diff
@@ -2251,8 +2265,20 @@ private static async Task ExecuteResultWriteResponse(IResult? result, HttpContex
         await EnsureRequestResultNotNull(result).ExecuteAsync(httpContext);
     }
 
-    private static Task WriteJsonResponse(HttpResponse response, object? value, JsonSerializerOptions? options)
+    private static Task WriteJsonResponse<T>(HttpResponse response, T? value, JsonSerializerOptions? options, JsonTypeInfo? jsonTypeInfo)
     {
+        if (jsonTypeInfo is not null &&
+            (jsonTypeInfo.PolymorphismOptions is not null ||
+            jsonTypeInfo.Type.IsValueType ||
+            jsonTypeInfo.Type == value?.GetType()))
+        {
+            // In this case the polymorphism is not
+            // relevant for us and will be handled by STJ, if needed.
+            return HttpResponseJsonExtensions.WriteAsJsonAsync(response, value!, (JsonTypeInfo<T>)jsonTypeInfo, default);
```

Comment:
❤️

For sealed types, value types and types with explicit polymorphism options, could we call this directly from the expression tree to avoid the if statement? I know even the if statement is pretty inexpensive here.

Reply:
Let me check what I can do